namerd/storage/zk/src/main/scala/com/twitter/finagle/serverset2/buoyant/ZkDtabStore.scala

```diff
@@ -0,0 +1,127 @@
+import com.twitter.finagle.serverset2.{RetryStream, Zk2Resolver, ZkSession => FZkSession}
+import com.twitter.finagle.stats.DefaultStatsReceiver
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.io.Buf
+import com.twitter.logging.Logger
+import com.twitter.util._
+import io.buoyant.namerd.DtabStore.{DtabNamespaceAlreadyExistsException, DtabNamespaceDoesNotExist, DtabVersionMismatchException}
+import io.buoyant.namerd.{DtabStore, VersionedDtab}
+import java.nio.ByteBuffer
+
+/**
+ * A DtabStore which stores dtabs in ZooKeeper.  Dtabs are stored, Utf8
+ * encoded, in the ZooKeeper node /dtabs/<ns>.
+ */
+class ZkDtabStore(
+  hosts: String,
+  zkPrefix: String,
+  sessionTimeout: Option[Duration]
+) extends DtabStore {
+
+  private[this] val log = Logger()
+  private[this] val retryStream = RetryStream()
+  private[this] val stats = DefaultStatsReceiver.scope("zkclient").scope(Zk2Resolver.statsOf(hosts))
+
+  private[this] implicit val timer = DefaultTimer.twitter
+  private[this] val builder = ClientBuilder()
+    .hosts(hosts)
+    .sessionTimeout(sessionTimeout.getOrElse(FZkSession.DefaultSessionTimeout))
+    .statsReceiver(stats)
+
+  private[this] val zkSession = new ZkSession(
+    retryStream,
+    retryStream,
+    () => builder.writer(),
+    stats
+  )
+  private[this] def actOf[T](go: ZooKeeperRW => Future[Watched[T]]): Activity[T] =
+    zkSession.watchedOperation(go(zkSession.zk))
+
+  private[this] val actNs = actOf(_.getChildrenWatch(zkPrefix)).map(_.children.toSet)
+
+  def list(): Future[Set[String]] = actNs.values.toFuture.flatMap(Future.const)
+
+  def create(ns: String, dtab: Dtab): Future[Unit] = {
+    val path = s"$zkPrefix/$ns"
+    log.info(s"Attempting to create dtab at $path")
+
+    zkSession.zk.create(
+      path,
+      Some(Buf.Utf8(dtab.show)),
+      Seq(Data.ACL.AnyoneAllUnsafe),
+      CreateMode.Persistent
+    ).rescue {
+        case KeeperException.NodeExists(_) =>
+          Future.exception(new DtabNamespaceAlreadyExistsException(ns))
+      }.unit
+  }
+
+  override def update(ns: String, dtab: Dtab, version: Buf): Future[Unit] = {
+    val path = s"$zkPrefix/$ns"
+    log.info(s"Attempting to update dtab at $path")
+
+    zkSession.zk.setData(
+      path,
+      Some(Buf.Utf8(dtab.show)),
+      Some(versionInt(version))
+    ).rescue {
+        case BadVersion(_) => Future.exception(new DtabVersionMismatchException)
+        case NoNode(_) => Future.exception(new DtabNamespaceDoesNotExist(ns))
+      }.unit
+  }
+
+  private[this] val dtabWatchOp = Memoize { path: String =>
+    log.info(s"Attempting to obverve $path")
+    actOf(_.getDataWatch(path)).flatMap { data =>
+      data.data match {
+        case Some(Buf.Utf8(s)) =>
+          Activity.value(Some(VersionedDtab(Dtab.read(s), versionBuf(data.stat.version))))
+        case None =>
+          Activity.exception(new IllegalStateException(s"Empty node: $path"))
+      }
+    }.handle {
+      case NoNode(_) => None
+    }
+  }
+
```

Comment:
Not a blocker, but this isn't great for something like namerd.  It means that once a node is observed, it will be observed forever (even if there are no observers).  If for instance, many short-lived dtabs are created and deleted, this will manifest as a memory leak.